platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -396,23 +396,31 @@ public Configuration setExtendsFrom(Iterable<Configuration> extendsFrom) {
     @Override
     public Configuration extendsFrom(Configuration... extendsFrom) {
         validateMutation(MutationType.HIERARCHY);
-        for (Configuration configuration : extendsFrom) {
-            if (configuration.getHierarchy().contains(this)) {
+        for (Configuration extended : extendsFrom) {
+            DefaultConfiguration other = Objects.requireNonNull(Cast.uncheckedCast(extended));
+            if (!Objects.equals(this.identityPath.getParent(), other.identityPath.getParent())) {
```

Comment:
Instead of decomposing the identity path of this configuration to get the project's identity path, we should get the project's identity path from the source (DomainObjectContext) instead. Each configuration already has a reference to its owning `DomainObjectContext`.

The DomainObjectContext however only exposes the path of the project within the build tree (`getProjectPath`) method. 

We should update `DomainObjectContext` to expose a `getProjectIdentityPath` and use that here instead

Reply:
https://github.com/gradle/gradle/issues/29893

Reply:
Greatly is a bit of an overstatement, and I'm not sure we really need extensive testing. 

I've implemented that issue, and the changes were minimal

https://github.com/gradle/gradle/pull/29894

Feel free to pull these changes into this PR, or we can merge my PR after and do additional work to integrate them.

But, I do to think this is the proper way to do it

Reply:
It looks correct, but it does already double the number of files, and that's without test coverage.  I think test coverage is vital for a change like that, to make sure it stays aligned with the other getters in `DefaultProject`, so I wouldn't want to rush it in without.

Updating this PR to use those methods after is trivial, let's just do that later.